### ISSUE TITLE
Preserve formatting tokens on render

### ISSUE THREAD

**Issue by jsorge** [iwasrobbed/Down#225]
Please help prevent duplicate requests before submitting a new one:

* [x] I've searched other open/closed issues for duplicates before opening up this new issue.

# Feature Request

If I have a markdown string such as:

```
This is some **really** great text I've got here!
```

When I render it as an attributed string using Down, the `**` characters are removed and the styling is applied to the text inside the pair of `**` tokens. I'd like to be able to preserve the tokens and display those in the attributed string, applying the styling to the tokens as well as the text they wrap.

## Is your feature request related to a problem? Please describe.

I'm frustrated that I lose changes to the original markdown syntax that's rendered in my text view, so changes that I make to the text don't easily fit back to the document I'm modifying.

## Describe the solution you'd like

I want to have the ability to use a text view (`DownView` or otherwise) and have it be editable, so that I can track the original source of my markdown text and apply changes using the standard text view delegate callbacks.

## Describe alternatives you've considered

I attempted to pair Craig Hockenberry's [MarkdownAttributedString](https://github.com/chockenberry/MarkdownAttributedString) library with Down to see if it could render back the attributed string but alas it could not on my initial try at it.

**Comment by johnxnguyen** [comment]
Hi @jsorge 👋  As @iwasrobbed-ks already mentioned, you could reinsert the tokens in your `AttributedStringVisitor` overrides. If you want the tokens to be styled the same as the content, just insert the tokens before calling the `super` implementation. This approach is simple and clean but it may not be what you're looking for.

> I want to have the ability to use a text view (DownView or otherwise) and have it be editable, so that I can track the original source of my markdown text and apply changes using the standard text view delegate callbacks.

If you want to have an editor with markdown capability, then I'm afraid there isn't yet a good solution for this that involves `Down`. I don't think `cmark` was designed to preserve the original input string, since it appears some crucial information is lost during the parsing, such as how many line breaks there are between paragraphs. (I could be wrong, there may be some API in `cmark` that I'm not aware of). This means that there may always be some discrepancy between the input and output strings. This alone makes it not suitable for use in a user editable text view, because the user would expect only the styling to change and not the characters.

In short, if you only want to preserve some inline tokens, then subclassing the visitor and reinserting the tokens yourself is the best way to go. If you need to preserve the characters of the input string after styling, then I'm afraid that we don't have a solution for this.

**Comment by iwasrobbed-ks** [comment]
Going to close this one out as it diverges away from Down's primary use case and there are also a number of workarounds available for this edge case. Feel free to reopen or comment here if you have additional concerns
